src/core/thread/dua_manager.cpp

```diff
@@ -528,9 +529,11 @@ void DuaManager::HandleDuaResponse(Coap::Message &aMessage, const Ip6::MessageIn
         ExitNow(error = aResult);
     }
 
-    VerifyOrExit(aResult == OT_ERROR_NONE && aMessage.GetCode() == Coap::kCodeChanged, error = OT_ERROR_PARSE);
+    VerifyOrExit(aResult == OT_ERROR_NONE, error = OT_ERROR_PARSE);
+    VerifyOrExit(aMessage.GetCode() == Coap::kCodeChanged || aMessage.GetCode() >= Coap::kCodeBadRequest,
+                 error = OT_ERROR_PARSE);
 
-    error = ProcessDuaResponse(aMessage);
+    error = ProcessDuaResponse(aMessage, &mRegisteringDua);
```

Comment:
It might not be necessary to pass `mRegisteringDua`?

Reply:
It seems not necessary in most cases. 

In the current implementation, `mRegisteringDua` is used in the `ProcessDuaResponse` method only when the COAP response code is not `2.04`. Most of time `ProcessDuaResponse` handles `DUA.rsp` in such case `mRegisteringDua` must be the DUA that is registering. 

But note that `ProcessDuaResponse` is also called by `HandleDuaNotification` which handles proactive DUA notification (`n/dn`). For `n/dn` message with code != 2.04, even though very unlikely to receive, we can't assume that the corresponding DUA is `mRegisteringDua`. 

Thoughts?

Reply:
I realized that the COAP code for `n/dn` must be POST (0.02). So, there is a bug int the current code. Will fix.

Reply:
fixed, please check